### PR TITLE
Rework

### LINKED ISSUE #5: Add tests

### LINKED ISSUE #6: Fix live github pages demo

### PR DESCRIPTION
**Pull request by kumavis**
Significant rewrite of LavaTube improving API and value access

- new API
```js
import LavaTube from '@lavamoat/lavatube'

LavaTube.iterate(start, opts)
LavaTube.walk(start, visitorFn, opts)
LavaTube.getAllValues(start, opts)
LavaTube.find(start, target, opts)
```

- expanded property walking to include Symbols
- no longer maintain an "ignores" list
- rewrite as esm
- drop bundling requirement for demos
- rewrite to use recursive iterators to avoid potential stack overflow
- remove unsupported options
- adds special case (with optional override) for values with known accessible internal slots like `Map` and `Set`
- walks breadth-first for improved performance
- optional weakmap lookups `shouldBruteForceWeakMaps`
- optional feature to call functions `shouldCallFunctions`
- optional feature to construct functions `shouldConstructFunctions`
- optional `depthFirstSearch`
- rewrite to track depth vs static maxDepth
- documents limitations and getter side effects
- simplify default generateKey
- expose options for cross-realm searches `globalThese`
- handle iterables and iterators

Fixes #5
Fixes #6 
Fixes #16 
Fixes #17

### CODE REVIEW

File: demo/tree/tree.js [hunk 1]
```diff
@@ -18,8 +20,16 @@ function tree(src, dst, limit) {
     }
 
     const tree = {};
-    new LavaTube(cb, {generateKey: key, maxRecursionLimit: limit}).walk(src);
+    new LavaTube({
```
Comment (weizman) [inline demo/tree/tree.js:23]: I'm still wondering what's the best API for this..

Should we copy off of [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver#example) maybe?

```javascript
mo = new MutationObserver(cb);
mo.observe(target, config);
```

```jaavscript
lt = new LavaTube(cb);
lt.walk(target, config);
```

Worth considering

Comment (kumavis) [reply demo/tree/tree.js]: redid the API

```javascript
import LavaTube from '@lavamoat/lavatube';

// get starting and target references for your search
const target = window;
const div = document.createElement('div');
document.body.appendChild(div);
const startRef = div;

// use "find" when looking for a path to a specific value
const path = LavaTube.find(startRef, target);

// use "walk" to visit every value by specifying a visitor function
// the visitor can return "true" to stop iteration and select the
// current value and path as the final result
const result = LavaTube.walk(startRef, checkValueForTarget);
if (result !== undefined) {
  const { value, path } = result;
  console.log(`found ${value} at ${path}`)
}

// use "iterate" to get an iterator for visiting values
for (const [value, path] of LavaTube.iterate(startRef)) {
  if (checkValueForTarget(value, path)) {
    console.log(`found ${value} at ${path}`)
    break;
  }
}

// configuration options can be specified as well
const opts = { maxDepth: 9 }
LavaTube.walk(window, (value, path) => {
  return checkValueForTarget(value, path);
}, opts);

function checkValueForTarget (value, path) {
  if (value === target) {
    console.log(`found ${value} at ${path}`)
    return true;
  }
  return false;
}

```

**Review by weizman — APPROVED** [review]
LGTM (except for it jamming [ProtoTree](https://weizmangal.com/ProtoTree/?filters=XML) completely when little to no filters applied)

**Comment by kumavis** [comment]
validated the ProtoTree issue is not directly related to LavaTube

**Pull request merged**